sources.sh

```diff
@@ -39,60 +54,145 @@ for tag in $externalPins; do
+								"lastStageFrom": '"$([ -n "$doHeavy" ] && echo -n '{{ if eq $builder "oci-import" -}}
+									{{- /* TODO remove this special case: https://github.com/docker-library/bashbrew/pull/92 */ -}}
+									"scratch"
+								{{- else -}}
+									{{ $.ArchLastStageFrom $a . | json }}
+								{{- end }}' || echo -n 'null')"',
+								"platformString": {{ (ociPlatform $a).String | json }},
+								"platform": {{ ociPlatform $a | json }},
+								"parents": { }
+							}
 						}
 					}
-				}
+				{{- end -}}
 			{{- end -}}
 		{{- end -}}
-	{{- end -}}
-' "$@" | jq 3>&1 1>&2 2>&3- -r '
-	# https://github.com/jqlang/jq/issues/2063 - "stderr" cannot functionally output a string correctly until jq 1.7+ (which is very very recent), so we hack around it to get some progress output by using Bash to swap stdout and stderr so we can output our objects to stderr and our progress text to stdout and "fix it in post"
-	# TODO balk / error at multiple arches entries
-	first(.arches | keys_unsorted[]) as $arch
-	| .arches[$arch].tags[0] as $tag
-	| stderr
-	| "\($tag) (\($arch)): \(.sourceId)"
-	# TODO if we could get jq 1.7+ for sure, we can drop this entire "jq" invocation and instead have the reduce loop of the following invocation print status strings directly to "stderr"
-' | jq -n --argjson pins "$externalPinsJson" '
+	' "$@" )
+	if [ -n "$doHeavy" ]; then
+		"${bbCat[@]}" | jq 3>&1 1>&2 2>&3- -r '
+			# https://github.com/jqlang/jq/issues/2063 - "stderr" cannot functionally output a string correctly until jq 1.7+ (which is very very recent), so we hack around it to get some progress output by using Bash to swap stdout and stderr so we can output our objects to stderr and our progress text to stdout and "fix it in post"
+			# TODO balk / error at multiple arches entries
+			first(.arches | keys_unsorted[]) as $arch
+			| .arches[$arch].tags[0] as $tag
+			| stderr
+			| "\($tag) (\($arch)): \(.sourceId)"
+			# TODO if we could get jq 1.7+ for sure, we can drop this entire "jq" invocation and instead have the reduce loop of the following invocation print status strings directly to "stderr"
+		' | jq -n '[ inputs ]'
+	else
+		"${bbCat[@]}" | jq -n '[ inputs ]'
+	fi
+}
+
+# merges heavy-to-calculate data from the second json input (list or map of sources) into the first json input (list of sources)
+# uses "mostlyUniqueBitsSum" as a rough analogue for sourceId to correlate data between the input lists
+#  (sourceId, reproducibleGitChecksum, SOURCE_DATE_EPOCH, froms, lastStageFrom)
+# echo '[{}, {},...] [{extraData},...]' | mergeData
+mergeData() {
+	jq --slurp '
+		def mostlyUniqueBitsSum($archKey):
+			.GitCommit
+			+ .Directory
+			+ .File
+			+ .Builder
+			+ $archKey
+			| @base64
+		;
+		(
+			.[1]
+			| map(
+				. as $source
+				| .arches
+				| keys
+				| map(
+					. as $a
+					| $source.entries[]
+					| {
+						(. | mostlyUniqueBitsSum($a)): { entry: ., source: $source }
+					}
+				)
+			) | add | add
```

Comment:
Instead of the double `map` and double `add`, we should construct an array and then use `from_entries` (which can probably make everything less deeply nested):

```jq
[
	.[1][] as $source
	| $source.arches
	| keys[] as $arch
	| $source.entries[]
	| {
		key: mostlyUniqueBitsSum($arch),
		value: {
			entry: .,
			source: $source,
		},
	}
] | from_entries
```

(I think that could even be `mostlyUniqueBitsSum($source.arches | keys[])`, but that's probably taking things a little bit _too_ far since the multiplicative nature of that is less obvious and harder to control/maintain.)